src/webui/btjson.cpp

```diff
@@ -454,21 +506,214 @@ QByteArray btjson::getFilesForTorrent(const QString& hash)
+    ret[KEY_TORRENT_UPSPEED] = status.upload_payload_rate;
+    if (QBtSession::instance()->isQueueingEnabled() && h.queue_position(status) >= 0)
+        ret[KEY_TORRENT_PRIORITY] = h.queue_position(status);
+    else
+        ret[KEY_TORRENT_PRIORITY] = -1;
+    ret[KEY_TORRENT_SEEDS] = status.num_seeds;
+    ret[KEY_TORRENT_NUM_COMPLETE] = status.num_complete;
+    ret[KEY_TORRENT_LEECHS] = status.num_peers - status.num_seeds;
+    ret[KEY_TORRENT_NUM_INCOMPLETE] = status.num_incomplete;
+    const qreal ratio = QBtSession::instance()->getRealRatio(status);
+    ret[KEY_TORRENT_RATIO] = (ratio > QBtSession::MAX_RATIO) ? -1 : ratio;
+    ret[KEY_TORRENT_STATE] = h.torrentState().toString();
+    ret[KEY_TORRENT_ETA] = h.eta();
+    ret[KEY_TORRENT_SEQUENTIAL_DOWNLOAD] = status.sequential_download;
+    if (h.has_metadata())
+        ret[KEY_TORRENT_FIRST_LAST_PIECE_PRIO] = h.first_last_piece_first();
+    ret[KEY_TORRENT_LABEL] = TorrentPersistentData::getLabel(h.hash());
+
+    return ret;
+}
+
+// Compare two structures (prevData, data) and calculate difference (syncData).
+// Structures encoded as map.
+void processMap(QVariantMap prevData, QVariantMap data, QVariantMap &syncData)
+{
+    // initialize output variable
+    syncData.clear();
+
+    QVariantList removedItems;
+    foreach (QString key, data.keys()) {
+        removedItems.clear();
+
+        switch (data[key].type()) {
+        case QVariant::Map: {
+                QVariantMap map;
+                processMap(prevData[key].toMap(), data[key].toMap(), map);
+                if (!map.isEmpty())
+                    syncData[key] = map;
+            }
+            break;
+        case QVariant::Hash: {
+                QVariantHash hash;
+                processHash(prevData[key].toHash(), data[key].toHash(), hash, removedItems);
+                if (!hash.isEmpty())
+                    syncData[key] = hash;
+                if (!removedItems.isEmpty())
+                    syncData[key + KEY_SUFFIX_REMOVED] = removedItems;
+            }
+            break;
+        case QVariant::List: {
+                QVariantList list;
+                processList(prevData[key].toList(), data[key].toList(), list, removedItems);
+                if (!list.isEmpty())
+                    syncData[key] = list;
+                if (!removedItems.isEmpty())
+                    syncData[key + KEY_SUFFIX_REMOVED] = removedItems;
+            }
+            break;
+        case QVariant::String:
+        case QVariant::LongLong:
+        case QMetaType::Float:
+        case QVariant::Int:
+        case QVariant::Bool:
+        case QVariant::Double:
+        case QVariant::ULongLong:
+            if (prevData[key] != data[key])
+                syncData[key] = data[key];
+            break;
+        default:
+            Q_ASSERT(0);
+        }
+    }
+}
+
+// Compare two lists of structures (prevData, data) and calculate difference (syncData, removedItems).
+// Structures encoded as map.
```

Comment:
This is because, apparently, that we discussed some things in Russian. 'll take a small piece:
>The main idea here is that we should not operate with some abstract QVariantList, QVariantHash and QVariantMap - we've got just a way of representing, respectively, the list of simple data, list of structured data and the structured data itself.